opengovsg/postmangovsg
# Change csv upload complete from waiting to polling

## Problem

Currently, a response is only returned after the uploaded csv is done processing and written to the db. This results in a timeout if processing is too long (i.e. csv is too large). A temporary fix was deployed previously to increase the timeouts of axios and nginx on the alb. 

Closes #55
Closes #323

## Solution

`/upload/complete` to return 202 on download from s3 bucket and hydrating first message. The bulk insert to message table which takes up majority of the processing time will complete asynchronously. 

Frontend will poll for status of csv processing.

How it works:
- after user upload csv to s3 and calls `/upload/complete`, server downloads file, attempts hydration on first row and checks for invalid recipient. If there are no errors, it will write the new filename to `s3_object` in `campaigns` table and return a 202.
- backend will poll for changes in csv processing status. During this time, user is not allowed to edit the campaign and is only restricted to the upload recipients tab.
- Backend will perform the bulk insert into db. If there are errors, it will rollback any changes and write the error into the `s3_object`. If not, it will replace the s3_object with the new filename and other metadata.
- If processing job exceeds 10 mins, it will be considered as timed out. This is to prevent the campaign from being stuck as uneditable if server dies while processing.

**Improvements**:

Frontend:
- Move upload csv frontend display logic to a common component `CsvUpload`
- Move frontend upload logic to shared `upload.service.ts`
- Some refactoring of types

Backend
- Move `s3_object` helper functions from `campaign.service.ts` to `template.service.ts`
- CanEditCampaign middleware also checks if  csv is processing.
- Campaign details now contain `is_csv_processing` key

## Screenshots
**AFTER**:
Processing
![Screenshot 2020-06-08 at 2 32 08 AM](https://user-images.githubusercontent.com/10072985/83977133-b6e1d800-a930-11ea-82ba-71710c3f7a5b.png)
Errored
![Screenshot 2020-06-08 at 2 32 58 AM](https://user-images.githubusercontent.com/10072985/83977135-b9443200-a930-11ea-978d-5359fafe3320.png)


## Tests

1. Csv is processing
- Add `"temp_filename": "<any string>"` to `s3_object`. Check if frontend is campaign details return `is_csv_processing`: true and frontend polls for data
- Also check that edit operations like changing template returns forbidden.

2. Csv processing has errored
- Add `"temp_filename": "<any string>"` and `"error": "<any string>"` to s3_object and reload frontend
- Check error message appears 
- Closing error block should delete error string and temp filename in db

3. Csv processing has timed out
- Add `"temp_filename": "<any string>"` and set `last_updated` to more than 10 minutes ago
- Reload frontend, timeout message should show

4. End-to-end test
- Generate large csv here: https://gist.github.com/pregnantboy/e17f4fa6423463c391bc05cda60ba569
- Refreshing during processing should work

## Deploy Notes

**New environment variables**:

- `CSV_PROCESSING_TIMEOUT_IN_MS` : Maximum permissible time for csv processing before it is marked as timed out (defaults to 10 mins)
